JInsight logs an IOException instead of reporting the body of the error response from server

In cases of an error response, JInsight should log the response body of the error response from the server, instead of logging an exception

```
java.io.IOException: Server returned HTTP response code: 400 for URL: https://api.apptuit.ai/api/put?details
        at sun.reflect.GeneratedConstructorAccessor62.newInstance(Unknown Source)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1944)
        at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1939)
        at java.security.AccessController.doPrivileged(Native Method)
        at sun.net.www.protocol.http.HttpURLConnection.getChainedException(HttpURLConnection.java:1938)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1508)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492)
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:263)
        at ai.apptuit.metrics.client.ApptuitPutClient.put(ApptuitPutClient.java:103)
        at ai.apptuit.metrics.dropwizard.ApptuitReporter.lambda$report$9(ApptuitReporter.java:109)
        at com.codahale.metrics.Timer.time(Timer.java:116)
        at ai.apptuit.metrics.dropwizard.ApptuitReporter.report(ApptuitReporter.java:107)
        at com.codahale.metrics.ScheduledReporter.report(ScheduledReporter.java:235)
        at com.codahale.metrics.ScheduledReporter$1.run(ScheduledReporter.java:174)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: Server returned HTTP response code: 400 for URL: https://api.apptuit.ai/api/put?details
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1894)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492)
        at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:347)
        at ai.apptuit.metrics.client.ApptuitPutClient.put(ApptuitPutClient.java:101)
        ... 12 more
java.io.IOException: Server returned HTTP response code: 400 for URL: https://api.apptuit.ai/api/put?details
```

--- a/src/main/java/ai/apptuit/metrics/client/ApptuitPutClient.java
+++ b/src/main/java/ai/apptuit/metrics/client/ApptuitPutClient.java
@@ -83,8 +83,10 @@ public class ApptuitPutClient {
 
     DatapointsHttpEntity entity = new DatapointsHttpEntity(dataPoints, globalTags);
 
+    HttpURLConnection urlConnection;
+    int status;
     try {
-      HttpURLConnection urlConnection = (HttpURLConnection) apiEndPoint.openConnection();
+      urlConnection = (HttpURLConnection) apiEndPoint.openConnection();
       urlConnection.setConnectTimeout(CONNECT_TIMEOUT_MS);
       urlConnection.setReadTimeout(SOCKET_TIMEOUT_MS);
       urlConnection.setChunkedStreamingMode(0);
@@ -102,17 +104,31 @@ public class ApptuitPutClient {
       entity.writeTo(outputStream);
       outputStream.flush();
 
-      int status = urlConnection.getResponseCode();
+      status = urlConnection.getResponseCode();
       debug("-------------------" + status + "---------------------");
-      InputStream inputStr = urlConnection.getInputStream();
+    } catch (IOException e) {
+      //TODO: Return status to caller, so they can choose to retry etc
+      LOGGER.error("Error posting data", e);
+      return;
+    }
+
+    try {
+      InputStream inputStr;
+      if (status < HttpURLConnection.HTTP_BAD_REQUEST) {
+        inputStr = urlConnection.getInputStream();
+      } else {
+        /* error from server */
+        inputStr = urlConnection.getErrorStream();
+      }
+
       String encoding = urlConnection.getContentEncoding() == null ? "UTF-8"
           : urlConnection.getContentEncoding();
       String responseBody = consumeResponse(inputStr, Charset.forName(encoding));
       debug(responseBody);
     } catch (IOException e) {
-      //TODO: Return status to caller, so they can choose to retry etc
-      LOGGER.error("Error posting data", e);
+      LOGGER.error("Error draining response", e);
     }
+
   }
 
   private String consumeResponse(InputStream inputStr, Charset encoding) {

--- a/src/test/java/ai/apptuit/metrics/client/ApptuitPutClientTest.java
+++ b/src/test/java/ai/apptuit/metrics/client/ApptuitPutClientTest.java
@@ -248,6 +248,7 @@ public class ApptuitPutClientTest {
         default:
           response = SUCCESS_RESPONSE_BODY.getBytes();
           status = HttpURLConnection.HTTP_OK;
+          break;
       }
       exchange.sendResponseHeaders(status, response.length);
       exchange.getResponseBody().write(response);
@@ -260,7 +261,7 @@ public class ApptuitPutClientTest {
       if (rawQuery == null) {
         return HttpURLConnection.HTTP_OK;
       }
-      if (rawQuery.equals("status=400")) {
+      if ("status=400".equals(rawQuery)) {
         return HttpURLConnection.HTTP_BAD_REQUEST;
       }
       return HttpURLConnection.HTTP_OK;

--- a/src/test/java/ai/apptuit/metrics/client/ApptuitPutClientTest.java
+++ b/src/test/java/ai/apptuit/metrics/client/ApptuitPutClientTest.java
@@ -30,6 +30,8 @@ import java.io.PipedInputStream;
 import java.io.PipedOutputStream;
 import java.net.HttpURLConnection;
 import java.net.InetSocketAddress;
+import java.net.MalformedURLException;
+import java.net.URI;
 import java.net.URL;
 import java.util.ArrayList;
 import java.util.HashMap;
@@ -37,7 +39,11 @@ import java.util.List;
 import java.util.Map;
 import java.util.Scanner;
 import java.util.zip.GZIPInputStream;
+import org.json.simple.parser.ParseException;
+import org.junit.After;
+import org.junit.AfterClass;
 import org.junit.Before;
+import org.junit.BeforeClass;
 import org.junit.Test;
 
 /**
@@ -46,8 +52,25 @@ import org.junit.Test;
 
 public class ApptuitPutClientTest {
 
+  private static MockServer httpServer;
+
   private TagEncodedMetricName tagEncodedMetricName;
   private HashMap<String, String> globalTags;
+
+  private static String streamToString(InputStream inputStream) throws IOException {
+    return new Scanner(new GZIPInputStream(inputStream)).useDelimiter("\0").next();
+  }
+
+  @BeforeClass
+  public static void setUpClass() throws Exception {
+    httpServer = new MockServer();
+    httpServer.start();
+  }
+
+  @AfterClass
+  public static void tearDownClass() throws Exception {
+    httpServer.stop(0);
+  }
 
   @Before
   public void setUp() throws Exception {
@@ -58,7 +81,11 @@ public class ApptuitPutClientTest {
     globalTags.put("host", "rajiv");
     globalTags.put("env", "dev");
     globalTags.put("dev", "rajiv");
+  }
 
+  @After
+  public void tearDown() throws Exception {
+    httpServer.resetCapturedData();
   }
 
 
@@ -78,7 +105,6 @@ public class ApptuitPutClientTest {
       }
     }
   }
-
 
   @Test
   public void testEntityMarshallingWithGZIP() throws Exception {
@@ -101,50 +127,36 @@ public class ApptuitPutClientTest {
   }
 
   @Test
-  public void testPut() throws Exception {
+  public void testPut200() throws Exception {
+    testPut(200);
+  }
+
+  @Test
+  public void testPut400() throws Exception {
+    testPut(400);
+  }
+
+  private void testPut(int status) throws MalformedURLException, ParseException {
     //Util.enableHttpClientTracing();
 
     int numDataPoints = 10;
     ArrayList<DataPoint> dataPoints = createDataPoints(numDataPoints);
 
-    int port = 9797;
-    String path = "/api/endpoint";
-    String token = "MOCK_APPTUIT_TOKEN";
+    URL apiEndPoint = httpServer.getUrl(status);
+    ApptuitPutClient client = new ApptuitPutClient(MockServer.token, globalTags, apiEndPoint);
+    client.put(dataPoints);
 
-    HttpServer httpServer = HttpServer.create(new InetSocketAddress(port), 0);
-
-    List<HttpExchange> exchanges = new ArrayList<>();
-    List<String> requestBodies = new ArrayList<>();
-    httpServer.createContext(path, exchange -> {
-      exchanges.add(exchange);
-      requestBodies.add(streamToString(exchange.getRequestBody()));
-
-      byte[] response = "{\"success\":1,\"failed\":0,\"errors\":[]}".getBytes();
-      exchange.sendResponseHeaders(HttpURLConnection.HTTP_OK, response.length);
-      exchange.getResponseBody().write(response);
-      exchange.close();
-    });
-
-    httpServer.start();
-
-    try {
-
-      URL apiEndPoint = new URL("http://localhost:" + port + path);
-      ApptuitPutClient client = new ApptuitPutClient(token, globalTags, apiEndPoint);
-      client.put(dataPoints);
-
-    } finally {
-      httpServer.stop(0);
-    }
+    List<HttpExchange> exchanges = httpServer.getExchanges();
+    List<String> requestBodies = httpServer.getRequestBodies();
 
     HttpExchange exchange = exchanges.get(0);
     assertEquals("POST", exchange.getRequestMethod());
-    assertEquals(path, exchange.getRequestURI().getPath());
+    assertEquals(MockServer.path, exchange.getRequestURI().getPath());
 
     Headers headers = exchange.getRequestHeaders();
     assertEquals("gzip", headers.getFirst("Content-Encoding"));
     assertEquals("application/json", headers.getFirst("Content-Type"));
-    assertEquals("Bearer " + token, headers.getFirst("Authorization"));
+    assertEquals("Bearer " + MockServer.token, headers.getFirst("Authorization"));
 
     DataPoint[] unmarshalledDPs = Util.jsonToDataPoints(requestBodies.get(0));
 
@@ -152,10 +164,6 @@ public class ApptuitPutClientTest {
     for (int i = 0; i < numDataPoints; i++) {
       assertEquals(getExpectedDataPoint(dataPoints.get(i), globalTags), unmarshalledDPs[i]);
     }
-  }
-
-  private String streamToString(InputStream inputStream) throws IOException {
-    return new Scanner(new GZIPInputStream(inputStream)).useDelimiter("\0").next();
   }
 
   private ArrayList<DataPoint> createDataPoints(int numDataPoints) {
@@ -175,6 +183,88 @@ public class ApptuitPutClientTest {
     tags.putAll(globalTags);
     return new DataPoint(dataPoint.getMetric(), dataPoint.getTimestamp(), dataPoint.getValue(),
         tags);
+  }
+
+  private static class MockServer {
+
+    private static final int port = 9797;
+    private static final String path = "/api/endpoint";
+    private static final String token = "MOCK_APPTUIT_TOKEN";
+    private static final String SUCCESS_RESPONSE_BODY = "{\"success\":1,\"failed\":0,\"errors\":[]}";
+    private static final String STATUS400_RESPONSE_BODY = "{\"success\":223,\"failed\":2,\"errors\":[{\"datapoint\":{\"metric\":\"tomcat.requests.duration.mean\",\"timestamp\":1513650393,\"value\":\"NaN\",\"tags\":{\"method\":\"DELETE\",\"context\":\"ROOT\",\"host\":\"ade-instance.c.pivotal-canto-171605.internal\",\"env\":\"dev\",\"collector\":\"jinsight\",\"status\":\"200\"}},\"error\":\"Unable to parse value to a number\"},{\"datapoint\":{\"metric\":\"tomcat.requests.duration.mean\",\"timestamp\":1513650393,\"value\":\"NaN\",\"tags\":{\"method\":\"POST\",\"context\":\"ROOT\",\"host\":\"ade-instance.c.pivotal-canto-171605.internal\",\"env\":\"dev\",\"collector\":\"jinsight\",\"status\":\"200\"}},\"error\":\"Unable to parse value to a number\"}]}";
+
+    private HttpServer httpServer;
+    private List<HttpExchange> exchanges = new ArrayList<>();
+    private List<String> requestBodies = new ArrayList<>();
+
+    public MockServer() throws IOException {
+      httpServer = HttpServer.create(new InetSocketAddress(port), 0);
+      httpServer.createContext(path, this::handleExchange);
+    }
+
+    public void start() {
+      httpServer.start();
+    }
+
+    public void stop(int i) {
+      httpServer.stop(i);
+    }
+
+    private URL getUrl() throws MalformedURLException {
+      return getUrl(HttpURLConnection.HTTP_OK);
+    }
+
+    private URL getUrl(int code) throws MalformedURLException {
+      String url = "http://localhost:" + port + path;
+      if (code != 200) {
+        url += "?status=" + code;
+      }
+      return new URL(url);
+    }
+
+    public List<HttpExchange> getExchanges() {
+      return exchanges;
+    }
+
+    public List<String> getRequestBodies() {
+      return requestBodies;
+    }
+
+    public void resetCapturedData() {
+      exchanges.clear();
+      requestBodies.clear();
+    }
+
+    private void handleExchange(HttpExchange exchange) throws IOException {
+      exchanges.add(exchange);
+      requestBodies.add(streamToString(exchange.getRequestBody()));
+
+      int status = getResponseType(exchange);
+      byte[] response = null;
+      switch (status) {
+        case HttpURLConnection.HTTP_BAD_REQUEST:
+          response = STATUS400_RESPONSE_BODY.getBytes();
+          break;
+        default:
+          response = SUCCESS_RESPONSE_BODY.getBytes();
+          status = HttpURLConnection.HTTP_OK;
+      }
+      exchange.sendResponseHeaders(status, response.length);
+      exchange.getResponseBody().write(response);
+      exchange.close();
+    }
+
+    private int getResponseType(HttpExchange exchange) {
+      URI uri = exchange.getRequestURI();
+      String rawQuery = uri.getRawQuery();
+      if (rawQuery == null) {
+        return HttpURLConnection.HTTP_OK;
+      }
+      if (rawQuery.equals("status=400")) {
+        return HttpURLConnection.HTTP_BAD_REQUEST;
+      }
+      return HttpURLConnection.HTTP_OK;
+    }
   }
 
 }

